flink-connectors/flink-connector-elasticsearch-base/src/main/java/org/apache/flink/streaming/connectors/elasticsearch/ElasticsearchSinkBase.java

```diff
@@ -165,20 +286,36 @@ public void beforeBulk(long executionId, BulkRequest request) { }
 				@Override
 				public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
 					if (response.hasFailures()) {
-						for (BulkItemResponse itemResp : response.getItems()) {
-							Throwable failure = callBridge.extractFailureCauseFromBulkItemResponse(itemResp);
+						BulkItemResponse itemResponse;
+						Throwable failure;
+
+						for (int i = 0; i < response.getItems().length; i++) {
+							itemResponse = response.getItems()[i];
+							failure = callBridge.extractFailureCauseFromBulkItemResponse(itemResponse);
 							if (failure != null) {
-								LOG.error("Failed Elasticsearch item request: {}", failure.getMessage(), failure);
-								failureThrowable.compareAndSet(null, failure);
+								LOG.error("Failed Elasticsearch item request: {}", itemResponse.getFailureMessage(), failure);
+
+								if (failureHandler.onFailure(request.requests().get(i), failure, requestIndexer)) {
+									failureThrowable.compareAndSet(null, failure);
+								}
 							}
 						}
 					}
+
+					numPendingRequests.getAndAdd(-request.numberOfActions());
 				}
 
 				@Override
 				public void afterBulk(long executionId, BulkRequest request, Throwable failure) {
-					LOG.error("Failed Elasticsearch bulk request: {}", failure.getMessage(), failure);
-					failureThrowable.compareAndSet(null, failure);
+					LOG.error("Failed Elasticsearch bulk request: {}", failure.getMessage(), failure.getCause());
+
+					// whole bulk request failures are usually just temporary timeouts on
+					// the Elasticsearch side; simply retry all action requests in the bulk
+					for (ActionRequest action : request.requests()) {
+						requestIndexer.add(action);
+					}
+
+					numPendingRequests.getAndAdd(-request.numberOfActions());
```

Comment:
Lets say a bulk with 500 actions fails, so we re-add the bulk again, but subtract 500 actions from the pending requests.

Now the bulk succeeds and we subtract 500 actions again. Which would make the num pending requests negative? and void the at least once guarantees?

Am I overseeing something here?